common/containers/container.go

```diff
@@ -0,0 +1,296 @@
+}
+
+// aliasSet returns the alias to fully-qualified name mapping stored in the container.
+func (c *Container) aliasSet() map[string]string {
+	if c == nil || c.aliases == nil {
+		return noAliases
+	}
+	return c.aliases
+}
+
+// candidatesWithAlias returns the resolved candidates from the container with any applicable
+// alias appended on the end of the list.
+func (c *Container) candidatesWithAlias(candidates []string, name string) []string {
+	// If an alias exists for the name, ensure it is searched last.
+	alias, found := c.aliasSet()[name]
+	if found {
+		return append(candidates, alias)
+	}
+	return candidates
+}
+
+// ContainerOption specifies a functional configuration option for a Container.
+//
+// Note, ContainerOption implementations must be able to handle nil container inputs.
+type ContainerOption func(*Container) (*Container, error)
+
+// Abbrevs configures a set of simple names as abbreviations for fully-qualified names.
+//
+// An alias can be useful when working with variables, functions, and especially types from
+// multiple namespaces:
+//
+//    // CEL object construction
+//    qual.pkg.version.ObjTypeName{
+//       field: alt.container.ver.FieldTypeName{value: ...}
+//    }
+//
+// Only one qualified path may be used as CEL container, so at least one of these references is
+// a long qualified name within an otherwise short CEL program. Using the following abbreviations,
+// the program becomes much simpler:
+//
+//    // CEL Go option
+//    Abbrevs("qual.pkg.version.ObjTypeName", "alt.container.ver.FieldTypeName")
+//    // Simplified Object construction
+//    ObjTypeName{field: FieldTypeName{value: ...}}
+//
+// There are a few rules for the qualified names and the simple abbreviations generated from them:
+// - Qualified names must be dot-delimited, e.g. `package.subpkg.name`.
+// - The last element in the qualified name is the simple name used as the abbreviation.
+// - Abbreviations names must not collide with each other.
+// - The abbreviation name must not collide with the root-level container name.
+//
+// Abbrevs are distinct from container-based references in the following important ways:
+// - Containers follow C++ namespace resolution rules with searches from the most qualified name
+//   to the least qualified name.
+// - Container references within the CEL program may be relative, and are resolved to fully
+//   qualified names at either type-check time or program plan time, whichever comes first.
+// - Abbreviations must resolve to a fully-qualified name.
+// - Resolved abbreviations do not participate in namespace resolution.
+// - Resolved abbreviations are searched after container names, including container names in the
+//   global scope.
+//
+// If there is ever a case where an identifier could be in both the container and in the alias,
+// the container wins as the container will continue to evolve over time and the program must be
+// forward compatible with changes in the container.
```

Comment:
Don't we want this to go the other way for forward compatibility?  If we've written an expression with container `a.b.c` to use `x` as an abbrev for `z.y.x`, then we certainly don't want `x` to silently switch meaning if `a.b.c.x` or `a.b.x` or `a.x` is added to proto definitions somewhere. If we want to use that new `a.b.c.x`, we can use the fully-qualified name for it.  We want an error if "x" is in the global (unqualified) scope, as that's probably a declared variable.  On the other hand, if we've already written expressions relying on container path search to have `x` mean `a.b.x`, then it's the responsibility of the integrator to search for collisions before introducing a new abbreviation.  The integrator has control over their set of aliases, but might not have control over the proto definitions they include.

Reply:
The container is a higher-order concept which may not be (completely) within the developer's control; whereas the alias is within the developer's control. In the ideal case the container wins, but if you want to ensure the alias is picked, using the leading '.' will ensure the proper behavior.

Reply:
I'm confused - how is the container not in the developer's control? Do you mean the contents of the proto namespace? If so, that's the point: with non-alias names first, a reference to an alias can silently change underneath you, but with alias first, all shadowing is deliberate. Yes, a leading "." will work, but then maybe recommend that and show it in examples?

Reply:
I'll LGTM if you're really sure this is the way you'd like it to go, but it might be worth writing up some guidance, maybe in the integration guide, of what to check for when adding new aliases, and how to update your proto definitions when you have stored some compiled ASTs.

Reply:
The current behavior was what we discussed at the language council and is what's present in the design doc. Also, the examples showcase the use of a leading '.' for disambiguation. 

In theory, both `container` and `alias` are in the developer's control. If the alias can shadow a containerized name it becomes difficult to reason about the source of truth for the identifier. Does it come from the container or the alias? If the alias wins, then it can shadow and render unusable parts of the container. If the container wins, it can change the meaning of the alias, unless the alias has a leading '.'. Given that each direction could shadow the other and they're both developer owned, I'd prefer to keep the maximal set of symbols available for use.

Reply:
Per our offline conversation, if we are recommending recompilation of a proto to address AST generation errors, then a stored AST is not sufficient to preserve the program meaning in light of changes in the container structure. To preserve the the meaning between recompilations, the aliases must be resolved first, and disambiguation will require using the fully qualified names rather than abbreviated forms.